File: src/web/middleware/pydantic_models/action.py
from pydantic import BaseModel
from typing import Optional, Any


class Action(BaseModel):
    id: str
    name: str
    status: str
    parameters: dict[str, Any]
    subaction: Optional["Action"]

    @classmethod
    def from_json(cls, json_data: dict[str, Any]) -> "Action":
        return cls(
            id=json_data["id"],
            name=json_data["name"],
            status=json_data["status"],
            parameters=json_data["parameters"],
            subaction=(
                Action.from_json(json_data["subaction"])
                if json_data["subaction"]
                else None
            ),
        )

    def to_json(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "name": self.name,
            "status": self.status,
            "parameters": self.parameters,
            "subaction": self.subaction.to_json() if self.subaction else None,
        }

File: src/web/frontend/robot_apps_flutter/packages/middleware_api_utilities/lib/src/models/action.dart
class Action {
  Action({
    required this.id,
    required this.name,
    required this.status,
    required this.parameters,
    required this.subaction,
  });

  factory Action.fromJson(Map<String, dynamic> json) {
    return Action(
      id: json['id'] as String,
      name: json['name'] as String,
      status: json['status'] as String,
      parameters: json['parameters'] as Map<String, dynamic>,
      subaction: json['subaction'] != null ? Action.fromJson(json['subaction'] as Map<String, dynamic>) : null,
    );
  }

  final String id;
  final String name;
  final String status;
  final Map<String, dynamic> parameters;
  final Action? subaction;

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'status': status,
      'parameters': parameters,
      'subaction': subaction?.toJson(),
    };
  }
}
